src/XML/SignedElementTrait.php

```diff
@@ -0,0 +1,108 @@
+<?php
+
+declare(strict_types=1);
+
+namespace SimpleSAML\XMLSecurity\XML;
+
+use Exception;
+use SimpleSAML\Assert\Assert;
+use SimpleSAML\XMLSecurity\XML\ds\Signature;
+use SimpleSAML\XMLSecurity\XMLSecurityKey;
+
+/**
+ * Helper trait for processing signed elements.
+ *
+ * @package simplesamlphp/xml-security
+ */
+trait SignedElementTrait
+{
+    /**
+     * The signature of this element.
+     *
+     * @var \SimpleSAML\XMLSecurity\XML\ds\Signature $signature
+     */
+    protected Signature $signature;
+
+
+    /**
+     * Get the signature element of this object.
+     *
+     * @return \SimpleSAML\XMLSecurity\XML\ds\Signature
+     */
+    public function getSignature(): ?Signature
+    {
+        return $this->signature;
+    }
+
+
+    /**
+     * Initialize a signed element from XML.
+     *
+     * @param \SimpleSAML\XMLSecurity\XML\ds\Signature $signature The ds:Signature object
+     */
+    protected function setSignature(Signature $signature): void
+    {
+        $this->signature = $signature;
+    }
+
+
+    /**
+     * Validate this element against a public key.
+     *
+     * true is returned on success, false is returned if we don't have any
+     * signature we can validate. An exception is thrown if the signature
+     * validation fails.
+     *
+     * @param  \SimpleSAML\XMLSecurity\XMLSecurityKey $key The key we should check against.
+     * @return bool True on success, false when we don't have a signature.
+     * @throws \Exception
+     */
+    public function validate(XMLSecurityKey $key): bool
+    {
+        if ($this->signature === null) {
+            return false;
+        }
+
+        Assert::eq(
+            $key->getAlgorithm(),
+            $this->signature->getAlgorithm(),
+            'Algorithm provided in key does not match algorithm used in signature.'
+        );
+
+        // check the signature
+        $signer = $this->signature->getSigner();
+        if ($signer->verify($key) === 1) {
+            return true;
+        }
+
+        throw new Exception("Unable to validate Signature");
+    }
+
+
+    /**
+     * Retrieve certificates that sign this element.
+     *
+     * @return array Array with certificates.
+     * @throws \Exception if an error occurs while trying to extract the public key from a certificate.
+     */
+    public function getValidatingCertificates(): array
+    {
+        $ret = [];
+        foreach ($this->signature->getCertificates() as $cert) {
```

Comment:
I think this can be quite expensive. Would this be an alternative approach?

1. Require `validate()` before calling `getValidatingCertificates()`. If we call the latter first, raise an exception.
2. During `validate()`, if the signature validation succeeds with the given key, check if it corresponds to any of the certificates stored internally. Build a list of validating certificates with all where the passed key corresponds to them.
3. When calling `getValidatingCertificates()` after `validate()`, return the list we just built. If none of the certificates  in the signature correspond to the given key, or no certificates where found, we just return an empty list.